mono/mini/interp/transform.c

```diff
@@ -2966,6 +2984,22 @@ generate (MonoMethod *method, RuntimeMethod *rtm, unsigned char *is_bb_start, Mo
 				g_assert(klass->valuetype);
 				SET_SIMPLE_TYPE(td.sp - 1, STACK_TYPE_MP);
 				break;
+			case CEE_MONO_TLS: {
+				gint32 key = read32 (td.ip + 1);
+				td.ip += 5;
+				g_assert (key < TLS_KEY_NUM);
+				ADD_CODE (&td, MINT_MONO_TLS);
+				WRITE32 (&td, &key);
+				PUSH_SIMPLE_TYPE (&td, STACK_TYPE_MP);
+				break;
+			}
+			case CEE_MONO_ATOMIC_STORE_I4:
+				CHECK_STACK (&td, 2);
+				SIMPLE_OP (td, MINT_STIND_I4);
+				ADD_CODE (&td, MINT_MONO_MEMORY_BARRIER);
```

Comment:
This implementation is not quite correct. The strongest barrier this instruction supports, `MONO_MEMORY_BARRIER_SEQ`, would require a full barrier on both sides of the store.

Additionally, doing the store itself as regular C `*dst = src`, as opposed to using an atomic intrinsic, is prone to esoteric compiler optimizations even with memory barriers. The best thing to do here would be to just call `InterlockedWrite` from `utils/atomic.h` to implement this instruction.